### ISSUE TITLE
Controller input for Menu and Random does not work

### ISSUE THREAD

**Issue by RattleSN4K3** [UltraStar-Deluxe/USDX#242]
As of code changes from #240, the controller input which simulates keyboard input is not working anymore. This includes button presses for random song selection and opening the menu which are handled by character inputs (on key down) of [M] and [R] (the only ones which are currently mapped).

Due to the pull request being closed I gonna quote the reply from there here as well.

> Is there a reason why you removed the following code?
> ```
>            // TODO: hacky workaround for enabling keyboard simulation of controllers. use SDL2 new input handling SDL_StartTextInput and SDL_StopTextInput(
>            else if Event.key.keysym.unicode = 0 then
>            begin
>              s1 := SDL_GetScancodeName(Event.key.keysym.scancode);
>              if Length(s1) = 1 then KeyCharUnicode := Ord(s1[1])
>            end;
>```
>
>That specific code was required for simulating keyboard input from controllers (such as [M] for menu and [R] for random).
>
>If that code was the reason of the duplicated letters, I agree there should have been a double check, but removing that code and removing that feature shouldn't have been done.
>```
>            else if (Event.type_ <> SDL_TEXTINPUT) and (Event.key.keysym.unicode = 0) then
>```
>
>That code part specifically translates SLD key input (in form of TSDL_ScanCode converted from TSDL_KeyCode) to a KeyCharUnicode representation like how the rest of the USDX code is expecting.

Who ever did test the issue on duplicate letters and being able to reproduce the issues, could test by keeping that specific code part in with the small changes to the input handling. I am not sure if reverting or adding that double check is sufficient.

**Comment by basisbit** [comment]
Event.key.keysym.unicode is deprecated and does not work correctly any more in some builds of SDL 2.0.5. If SDL_GetScancodeName is required, then that should be done already before the keypress is simulated. Or are those keys actually KeyPresses without causing the SDL_CONTROLLER/JoypadBUTTONDOWN event?

**Comment by RattleSN4K3** [comment]
**tl;dr:** Unicode is not required. KeyDown/KeyUp needs to be handled correctly, which is now removed as of changes done by #240. Controller doesn't send "Text" and therefore the old method with the added  double check can be added back in. But it does require testing (_duplicate letters_).


The whole code part is handling KeyDown and TextInput. The code for KeyDown was removed by this change of #240. Keyboard input is simulated by KeyDown (and KeyUp) on the Joystick code, where the Unicode value is set to 0 to properly have distinct events and simulated keyboard input is not recognized as text input. The unicode value/property is not needed.

There might be better solution but as stated in code, it was a hacky (and most easily to accomplish solution) without changing much of other code parts.


```
function TJoyController.SimulateKeyboard(Key: TSDL_KeyCode; Pressed: boolean; NoMouseOverride: boolean): boolean;
var
  JoyEvent: TSDL_Event;
  TempName: string;
begin
  Result := true;

  Log.LogInfo(Format('Simulate Key: %s  Pressed: %d', [SDL_GetScancodeName(SDL_GetScancodeFromKey(Key)), integer(Pressed)]), 'TJoy.Simulate');

  // switch back to non-mouse mode (alias keyboard simulation)
  if not NoMouseOverride then MouseMode := false;

  JoyEvent := Default(TSDL_Event);
  JoyEvent.type_ := ifthen(Pressed, SDL_KEYDOWN, SDL_KEYUP);
  JoyEvent.key.keysym.sym := Key;
  JoyEvent.key.keysym.scancode := SDL_GetScancodeFromKey(Key);

  // always send empty/zero unicode char as workaround. Check UMain.CheckEvents
  JoyEvent.key.keysym.unicode := 0;

  SDL_PushEvent(@JoyEvent);
end;
```

The joystick code tries to translate the controller/joystick input to a mapped keyboard input (SDL TSDL_KeyCode). If such button is property mapped, SimulateKeyboard is called with the keycode (in form of TSDL_KeyCode). It pushes the events (as KeyDown/KeyUp) to the SDL stack. The rest is handled by the main code. As that main code expected proper unicode characters (for input like [M] or [R]), such KeyDown/KeyUp events simulated from JoyStick needed to have the raw keycode/scancode converted/translated to a unicode character (and then gets overridden).

The joystick code works modular as such, but the main code does not differentiate scan code input (hotkeys) and text input (for search, name etc.) property (both raw events are forward to the same logic of code). That's way I added the note about SDL_StartTextInput and SDL_StopTextInput in combination of TextEditing event (which can also solve various issues with text input). And the need of distinguishing both events was required. That was the now removed workaround.
